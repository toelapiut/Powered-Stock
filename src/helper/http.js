import axios from 'axios';

export const quandl = axios.create({
  baseURL: process.env.process.env.REACT_APP_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
    'Allow': 'GET, PUT, HEAD, OPTIONS'
  }
});

quandl.interceptors.request.use((config) => {
  config.params = config.params || {};
  config.params['key'] = process.env.REACT_APP_API_KEY;
  return config;
});
