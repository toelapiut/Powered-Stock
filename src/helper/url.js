export const url = {
  markets: 'markets',
  stocks: 'stocks/'
};