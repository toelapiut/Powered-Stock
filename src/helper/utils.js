import humps from 'humps';
import checkPropTypes from 'check-prop-types';
import {timeFormat} from 'd3-time-format';
import {bisector} from 'd3-array';
import moment from 'moment';

export const transpose = (rows, cols) => {
  /*
   * check if the row and col are empty and throw an error to
   * inform of the necessary requirements
   */
  if (typeof rows === 'undefined' || typeof cols === 'undefined') {
    throw new Error('row and col can not be undefined');
  }

  /*
   * throw error if the array isn't an array or col isn't an array
   */
  if (!Array.isArray(rows)) throw new Error('row must be  array');
  if (!Array.isArray(cols)) throw new Error('col must be  array');

  let dataset = [];

  for (let row of rows) {
    /*
     * check the data type before iterating through the matrix
     */
    if (!Array.isArray(row)) throw new Error('row must be a matrix row: [[a, b]]');
    let obj = {};
    row.forEach((item, index) => {
      obj[cols[index]['name']] = item;
    });
    dataset.push(obj);
  }

  return dataset;
};


export const transposeStock = (rows) => {
  let data = [];
  for (let row of rows) {
    let obj = {};
    const [date, close] = row;
    obj['date'] = date;
    obj['close'] = close;
    data.push(obj);
  }
  return data;
};

export const toCamelCase = (item) => {
  if (typeof item === 'undefined') {
    return item;
  }
  return humps.camelizeKeys(item);
};

export const checkPositivity = (num) => {
  if (typeof num === 'undefined') throw new Error('argument must not be empty');
  if (typeof num !== 'number') throw new Error('argument is not a number');
  let check = Math.sign(Number(num));
  return check !== -1;
};

export const formatDate = timeFormat('%b %d, \'%y');

export const getDate = (d) => new Date(d.date);

export const getStockValue = (d) => d.close;

export const bisectDate = bisector(d => new Date(d.date)).left;

export const getWindowDimensions = () => {
  const {innerWidth: width, innerHeight: height} = window;
  return {
    width,
    height
  };
};

export const dateFormat = (date, format = 'DD MMM, YYYY') => {
  return moment(date).format(format);
};

export const findTestByAttr = (component, attr) => {
  return component.find(`[data-test='${attr}']`);
};

export const checkProps = (component, expectedProps) => {
  checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

