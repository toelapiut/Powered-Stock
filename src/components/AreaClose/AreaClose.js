import React, {useMemo, useCallback} from 'react';
import {AreaClosed, Line, Bar} from '@visx/shape';
import appleStock from '@visx/mock-data/lib/mocks/appleStock';
import {curveMonotoneX} from '@visx/curve';
import {GridRows, GridColumns} from '@visx/grid';
import {scaleTime, scaleLinear} from '@visx/scale';
import {Tooltip, TooltipWithBounds, defaultStyles} from '@visx/tooltip';
import {localPoint} from '@visx/event';
import {LinearGradient} from '@visx/gradient';
import {max, extent} from 'd3-array';
import {formatDate, getDate, getStockValue, bisectDate} from '../../helper/utils';
import PropTypes from 'prop-types';


const stock = appleStock.slice(120);
export const background = '#222352';
export const background2 = '#222352';
export const accentColor = '#edffea';
export const accentColorDark = '#75daad';

import {getWindowDimensions} from '../../helper/utils';

let dimensions = getWindowDimensions();

console.log({dimensions});

const tooltipStyles = {
  ...defaultStyles,
  background,
  border: '1px solid white',
  color: 'white',
};

export const AreaClose = ({
  width = dimensions.width - 470,
  height = dimensions.height- 190,
  margin = {top: 0, right: 0, bottom: 0, left: 0},
  showTooltip,
  hideTooltip,
  tooltipData,
  tooltipTop = 0,
  tooltipLeft = 0,
}) => {
  if (width < 10) return null;

  // bounds
  const innerWidth = width - margin.left - margin.right;
  const innerHeight = height - margin.top - margin.bottom;
  // scales
  const dateScale = useMemo(
    () =>
      scaleTime({
        range: [margin.left, innerWidth + margin.left],
        domain: extent(stock, getDate),
      }),
    [innerWidth, margin.left],
  );

  const stockValueScale = useMemo(
    () =>
      scaleLinear({
        range: [innerHeight + margin.top, margin.top],
        domain: [0, (max(stock, getStockValue) || 0) + innerHeight / 3],
        nice: true,
      }),
    [margin.top, innerHeight],
  );

  // tooltip handler
  const handleTooltip = useCallback(
    (event) => {
      const {x} = localPoint(event) || {x: 0};
      const x0 = dateScale.invert(x);
      const index = bisectDate(stock, x0, 1);
      const d0 = stock[index - 1];
      const d1 = stock[index];
      let d = d0;

      if (d1 && getDate(d1)) {
        d = x0.valueOf() - getDate(d0).valueOf() > getDate(d1).valueOf() - x0.valueOf() ? d1 : d0;
      }

      showTooltip({
        tooltipData: d,
        tooltipLeft: x,
        tooltipTop: stockValueScale(getStockValue(d)),
      });
    },
    [showTooltip, stockValueScale, dateScale],
  );

  return (
    <div>
      <svg width={width} height={height}>
        <rect
          x={0}
          y={0}
          width={width}
          height={height}
          fill="url(#area-background-gradient)"
          rx={10}
        />
        <LinearGradient id="area-background-gradient" from={background} to={background2}/>
        <LinearGradient id="area-gradient" from={accentColor} to={accentColor} toOpacity={0.1}/>
        <GridRows
          left={margin.left}
          scale={stockValueScale}
          width={innerWidth}
          strokeDasharray="1,3"
          stroke={accentColor}
          strokeOpacity={0}
          pointerEvents="none"
        />
        <GridColumns
          top={margin.top}
          scale={dateScale}
          height={height}
          strokeDasharray="1,3"
          stroke={accentColor}
          strokeOpacity={0.2}
          pointerEvents="none"
        />
        <AreaClosed
          data={stock}
          x={d => dateScale(getDate(d)) ?? 0}
          y={d => stockValueScale(getStockValue(d)) ?? 0}
          yScale={stockValueScale}
          strokeWidth={1}
          stroke="url(#area-gradient)"
          fill="url(#area-gradient)"
          curve={curveMonotoneX}
        />
        <Bar
          x={margin.left}
          y={margin.top}
          width={innerWidth}
          height={innerHeight}
          fill="transparent"
          rx={14}
          onTouchStart={handleTooltip}
          onTouchMove={handleTooltip}
          onMouseMove={handleTooltip}
          onMouseLeave={() => hideTooltip()}
        />
        {tooltipData && (
          <g>
            <Line
              from={{x: tooltipLeft, y: margin.top}}
              to={{x: tooltipLeft, y: innerHeight + margin.top}}
              stroke={accentColorDark}
              strokeWidth={2}
              pointerEvents="none"
              strokeDasharray="5,2"
            />
            <circle
              cx={tooltipLeft}
              cy={tooltipTop + 1}
              r={4}
              fill="black"
              fillOpacity={0.1}
              stroke="black"
              strokeOpacity={0.1}
              strokeWidth={2}
              pointerEvents="none"
            />
            <circle
              cx={tooltipLeft}
              cy={tooltipTop}
              r={4}
              fill={accentColorDark}
              stroke="white"
              strokeWidth={2}
              pointerEvents="none"
            />
          </g>
        )}
      </svg>
      {tooltipData && (
        <div>
          <TooltipWithBounds
            key={Math.random()}
            top={tooltipTop - 12}
            left={tooltipLeft + 12}
            style={tooltipStyles}
          >
            {`$${getStockValue(tooltipData)}`}
          </TooltipWithBounds>
          <Tooltip
            top={innerHeight + margin.top - 14}
            left={tooltipLeft}
            style={{
              ...defaultStyles,
              minWidth: 72,
              textAlign: 'center',
              transform: 'translateX(-50%)',
            }}
          >
            {formatDate(getDate(tooltipData))}
          </Tooltip>
        </div>
      )}
    </div>
  );
};

AreaClose.propTypes = {
  width: PropTypes.number,
  height: PropTypes.number,
  margin: PropTypes.shape({
    top: PropTypes.number,
    right: PropTypes.number,
    bottom: PropTypes.number,
    left: PropTypes.number,
  }),
  showTooltip: PropTypes.func,
  hideTooltip: PropTypes.func,
  tooltipData: PropTypes.object,
  tooltipTop: PropTypes.number,
  tooltipLeft: PropTypes.number,
};

