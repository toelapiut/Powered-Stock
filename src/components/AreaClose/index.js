import {withTooltip} from '@visx/tooltip';
import {AreaClose} from './AreaClose';

export default withTooltip(AreaClose);