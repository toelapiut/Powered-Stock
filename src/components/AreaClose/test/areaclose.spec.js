import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {findTestByAttr, checkProps} from '../../../helper/utils';
import AreaClose from '../';


const setUp = (props = {}) => {
  return shallow(<AreaClose {...props}/>);
};

let props = {
  stock: [
    {
      'date': '1980-12-31',
      'close': 34.13
    },
    {
      'date': '1981-01-31',
      'close': 28.25
    },

    {
      'date': '2018-03-31',
      'close': 168.34
    }
  ],
  width: 970,
  tooltipTop: 0,
  tooltipLeft: 0,
  height: 710,
  hideTooltip: () => null,
  margin: {top: 0, right: 0, bottom: 0, left: 0},
  showTooltip: () => null,
  tooltipData: undefined,
};

describe('AreaClose Component', () => {
  describe('AreaClose renders tests', () => {
    let component;
    beforeAll(() => {
      component = setUp(props);
    });

    test('should render snapshot correctly', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });

  describe('Check property types', () => {
    it('should not throw a warning', () => {
      const propErr = checkProps(AreaClose, props);
      expect(propErr).toBeUndefined();
    });
  });

  describe('Check for width', () => {
    let component;
    let calendarProps = {
      ...props,
      width: 5
    };
    beforeAll(() => {
      component = setUp(calendarProps);
    });

    it('should return null if width is less 10', () => {
      const wrapper = findTestByAttr(component, 'area-close-container');
      expect(wrapper.length).toBe(0);
    });
  });

  describe('toggle ToolTip component', () => {
    let component;
    let calendarProps = {
      ...props,
      tooltipData: {date: '2017-08-31', close: 164}
    };
    beforeAll(() => {
      component = setUp(calendarProps);
    });

    it('should not render too', () => {
      const wrapper = findTestByAttr(component, 'toggle-tool-tip');
      expect(wrapper.length).toBe(0);
    });
  });
});