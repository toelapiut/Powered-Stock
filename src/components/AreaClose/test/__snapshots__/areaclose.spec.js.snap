// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AreaClose Component AreaClose renders tests should render snapshot correctly 1`] = `
<div
  style={
    Object {
      "height": "inherit",
      "position": "relative",
      "width": "inherit",
    }
  }
>
  <AreaClose
    height={710}
    hideTooltip={[Function]}
    margin={
      Object {
        "bottom": 0,
        "left": 0,
        "right": 0,
        "top": 0,
      }
    }
    showTooltip={[Function]}
    stock={
      Array [
        Object {
          "close": 34.13,
          "date": "1980-12-31",
        },
        Object {
          "close": 28.25,
          "date": "1981-01-31",
        },
        Object {
          "close": 168.34,
          "date": "2018-03-31",
        },
      ]
    }
    tooltipLeft={0}
    tooltipOpen={false}
    tooltipTop={0}
    updateTooltip={[Function]}
    width={970}
  />
</div>
`;
